### PR TITLE
Allow brute filter configuration

### PR DESCRIPTION
**Pull request by andrews05**
This PR allows the Brute filter parameters to be configured rather than using fixed constants. It also tweaks the presets to take advantage of these new options:
o3: level=1, lines=3 (fractionally worse/faster)
o4: level=1, lines=4 (same as before)
o5: level=4, lines=4 (fractionally better/slower)
o6: level=5, lines=8 (slightly better/slower)

The parameters can be explicitly set in both the API and the CLI.

### CODE REVIEW

**Comment by TPS** [comment]
> An open question is whether we should continue the existing pattern of short long-form options like `--bl` or do them properly like `--brute-lines`.

The great thing is, since you're _adding_ options, do whatever (all of the following obv imho):
 - The consistency part comes later, when _maintaining_ them: don't willy-nilly _change/remove_ them w/o providing some sort of backwards compatibility, to avoid breaking stuff downstream
- Short options for commonly used switches, long 1s for the more obscure.  That _should_ satisfy anyone.
 - _Especially_ on Windows terminals, _how_ to use each kind is constant, e.g., does 1 use a space, equals, colon, &c between a switch & parameter, & what delimiters are _required_ (none, quotes, double-quotes, &c).


Sorry for ranting.  Not experiencing this w/ `oxipng` per se, but I hate having to search through a tool's longhelp scanning for a ½-recalled option, just to go "how on earth was I supposed to remember _that_ ?!"

**Comment by andrews05** [comment]
> * The consistency part comes later, when _maintaining_ them: don't willy-nilly _change/remove_ them w/o providing some sort of backwards compatibility, to avoid breaking stuff downstream

Exactly, which is why it's important to get it right now so that we don't decide later that we made the wrong choice but can't change it until v11 due to being a BC break.

> * Short options for commonly used switches, long 1s for the more obscure.  That _should_ satisfy anyone.

These are definitely more obscure options. Proper long form* may indeed be the right choice.

> * _Especially_ on Windows terminals, _how_ to use each kind is constant, e.g., does 1 use a space, equals, colon, &c between a switch & parameter, & what delimiters are _required_ (none, quotes, double-quotes, &c).

I think we already cover this well enough 🙂

*I say "proper" long form to distinguish from some of our existing options like `--zc` which are technically long-form by virtue of being more than one character and requiring a double dash, but obviously aren't very long. I'm not sure this is good practice. The purpose of long form is to convey the meaning of the option right in the name.
That said, something like `--zopfli-iterations-without-improvement` seems a bit too much of a mouthful, so abbreviations like `--ziwi` may be appropriate in some cases.

**Comment by TPS** [comment]
> *I say "proper" long form to distinguish from some of our existing options like `--zc` which are technically long-form by virtue of being more than one character and requiring a double dash, but obviously aren't very long. I'm not sure this is good practice. The purpose of long form is to convey the meaning of the option right in the name.

This _is_ confusing to me & likely others.  AfaIk, `oxipng` is the only app that has double-dash short-forms.  Would it cause mayhem to _also_ accept single-dash for all short forms?

**Comment by andrews05** [comment]
This is the distinction:

Short form:
- Single character
- Single dash
- Can be combined together, e.g. `-s -a -v` can be `-sav`

Long form:
- Multiple characters
- Two dashes
- Cannot be combined

So despite being very "short", `--zc` and others like `--nx`, `--nz` etc are technically long form, and can't be made short form without reducing them to a single character. Many programs, especially older ones, do not follow this convention and just always use single dashes regardless, but this is not considered good practice today.
It might be better for e.g. `--nx` to be written as `--no-reductions`, but I'd be reluctant to make changes like that just for the sake of it.

**Comment by ace-dent** [comment]
Hey @andrews05 I got lost in the pedantry of naming here... rather than the bigger change! I don't really understand the intent here... 😅 

1. Currently what default settings are used for `brute`? Is it `zc=12` (highest libdeflate setting) and `lines=1`?
2. Is your lines setting to effectively skip rows? So test every other row? Or...?
3. If the intent is to increase speed, will you be performing benchmarking to make sure we don't regress  compression here? 

I'm a niche user who wants everything maxed out, with no regard for time... but .... Lower levels of libdeflate favour images that compress well with RLE / Huffman only, etc. ... It may be beneficial to run with a low and a high setting for libdeflate...?

**Comment by andrews05** [comment]
@ace-dent Since this is a draft, I didn't initially bother adding details to the top post (was going to add it later). I've now updated it with a little explainer. See also the long help for more info on the brute options.

I know libdeflate did have a problem where it wasn't trying Huffman-only strategies but that has since been resolved. I'm not aware of any remaining issues there?

File: src/main.rs [hunk 1]
```diff
@@ -205,6 +204,21 @@ fn parse_opts_into_struct(
...
+            if let Some(new_lines) = matches.get_one::<usize>("brute-lines") {
+                *num_lines = *new_lines;
+            }
+            if let Some(new_level) = matches.get_one::<i64>("brute-level") {
+                *level = *new_level as u8;
+            }
+        }
+        new_filters.insert(f);
+    }
+    opts.filter = new_filters;
```
Comment (andrews05) [inline src/main.rs:220]: This feels a bit awkward but I'm not sure if there's a better way to do it. Values in an index set aren't mutable like they are in a vec.

**Comment by ace-dent** [comment]
Thanks @andrews05.

> ... I know libdeflate did have a problem where it wasn't trying Huffman-only strategies ...

It's more a 'feature', where higher levels favour more complex compression strategies, which can prove less efficient in edge-cases. e.g. #638. Hence for exhaustive testing, I still test `--zc 1-12`.

**Comment by ace-dent** [comment]
I still don't really get these new parameters... hopefully it will all become apparent in time! 😅

**Comment by andrews05** [comment]
For each line, the brute method tries each filter and compresses the result to see which one is smallest. On it's own this isn't very effective though, so we include a few previous lines as context when compressing. The lines parameter is the total number of lines we compress at once.

**Comment by Winterhuman** [comment]
@andrews05 Quick question. When it says "Higher values are not necessarily better", does that mean using higher line or brute levels can yield worse compression than lower levels, or just that it won't yield anything better most of the time?

**Comment by TPS** [comment]
@Winterhuman We've already had examples of this raised, e.g. #638 → #639.  I believe that text is to clarify that unpredictable result.

**Comment by Winterhuman** [comment]
Ah, so that's true of these line & brute levels too, gotcha. Thanks for clarifying!

**Comment by andrews05** [comment]
So the Brute filter, like any of the other heuristic filters, is just an attempt to find filters that compress well. It works reasonably well, but a good outcome is not guaranteed.
From my own testing and observations, increasing the parameters can yield better results *on average*, but only to a point, and yes, it can be worse for some images. At `-omax` it's using level=5, lines=8. Beyond this you're essentially just adding random variance into the mix which may find a good result for one image, but not others. On one of my test sets I did find overall improvement with lines=16, but that didn't seem to hold true on other test sets. See also https://github.com/oxipng/oxipng/issues/730#issuecomment-3310133672 where I got a good result on one image with level=5, lines=6, but everything else I tried was worse.

You may like to experiment for yourself and see what you find, but at the end of the day these are advanced settings that most users shouldn't be touching.

[edit] One might assume that maxing it out with level=12, lines=&lt;height of your image&gt; would give the best result. This is not the case though, and to be honest I don't know why. (It would also be incredibly slow if you did this)

**Comment by ace-dent** [comment]
@andrews05 - Does this make sense for a brute force script? ...

```sh
oxipng --zopfli --zi 9000 --ziwi 300 --filters 0-8 "${png_file}"
oxipng --zopfli --zi 9000 --ziwi 300 --filters 9 --brute-level 1 --brute-lines 4 "${png_file}"
oxipng --zopfli --zi 9000 --ziwi 300 --filters 9 --brute-level 12 --brute-lines 8 "${png_file}"
```

Rationale: I want to test `Brute` filter with small and larger context windows (4 vs 8 lines).

**Comment by andrews05** [comment]
If you really want to brute force it you'll need to permute all combinations of levels and lines 😂
But if you're wanting to be more pragmatic and just try a few options, I'd probably stick with level 5 and just vary the lines, e.g. 4, 8, 12, 16, or even more if want to experiment.
In my own tests, I did not find that higher levels were beneficial. And as noted above, fiddling with the parameters is often just a matter of adding random variance that may or may not work well for any given image. There's no real science here unfortunately.

**Comment by Winterhuman** [comment]
(Currently working on all those permutations btw. Stay tuned)

**Comment by ace-dent** [comment]
Thanks @andrews05. And @Winterhuman I look forward to your results.
I *do* have a theory why higher line numbers may hinder compression... but need to check what the code is actually doing.

**Comment by Winterhuman** [comment]
The updated version of Sisyphus is now available with the new brute permutations (`--brute-level {1..12}` & `--brute-lines {2..16}`): https://gist.github.com/Winterhuman/21d7b148db40ff041f397b07a7aafb83

Be warned: up to 73000+ trials are created; a 96-byte favicon took nearly 3.5 minutes. The script tries to truncate & symlink identical & invalid trials where it can, but expect high CPU and memory (`--max-procs` can be specified to reduce or increase the time/computation cost; the default is 8).

**Comment by TPS** [comment]
_(@Winterhuman Did you see https://github.com/oxipng/oxipng/issues/658 for further "inspiration"?)_

**Comment by Winterhuman** [comment]
Actually, hold on, does `--brute-level`/`--brute-lines` actually do anything when `--zopfli` is specified? Just double checking in case I can save on permutations

**Comment by andrews05** [comment]
@Winterhuman Yes, but you probably want to do what @ace-dent was doing above, running filters 0-8 on their own and then permuting the brute options with only the brute filler. And feel free to go above 16 lines, I seem to recall one image got an improvement for me at around 128 lines. You probably want to cap it to the height of the image though.

**Comment by andrews05** [comment]
Also, you may be interested in checking out #749. It should remove the need to permute the -n flags.

**Comment by Winterhuman** [comment]
@andrews05 I've updated the script with your feedback!

1. The `--filters 0-8` and `--filters 9 --brute-level --brute-lines` variations are now separate.
2. The script now limits `--brute-lines` to the source image's height; and also, the script now has its own `-b|--brute-lines` option to set the maximum `oxipng --brute-lines` value to try up to (the default is still 16).
